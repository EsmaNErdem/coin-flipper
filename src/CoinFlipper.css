.CoinFlipper {
    width: 750px;
    height: 750px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    text-align: center;
}