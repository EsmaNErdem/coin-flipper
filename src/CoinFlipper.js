import React, {useState} from 'react';
import { getRandom } from './helpers';
import Coin from "./Coin"

const CoinFlipper = (props) => {
    const [coin, setCoin] = useState(null)
    const [headCounter, setHeadCounter] = useState(0)
    const [tailCounter, setTailCounter] = useState(0)

    const handleClick = () => {
        const flippedCoin = getRandom(props.coins)
        setCoin(flippedCoin)
        if(flippedCoin.side == "head") {
            setHeadCounter(headCounter + 1)
        } else {
            setTailCounter(tailCounter + 1)
        }
    }

    return (
        <div className='CoinFlipper'>
            <h1>Let's flip</h1>
            {coin && <Coin imgSrc={coin.imgSrc} side={coin.side} />}
            <button onClick={handleClick} data-testid="flipping-button">Flip it!</button>
            <p> Out of {headCounter + tailCounter}, head: {headCounter} and tail: {tailCounter}</p>
        </div>
    )
}

CoinFlipper.defaultProps = {
    coins: [
      {
        side: "head",
        imgSrc: "https://www.pcgs.com/UserImages/71009269o.jpg"
      },
      {
        side: "tail",
        imgSrc: "https://www.pcgs.com/UserImages/71009269r.jpg"
      }
    ]
};

export default CoinFlipper;