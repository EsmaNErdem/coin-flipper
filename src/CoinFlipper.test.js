import React from "react";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import CoinFlipper from "./CoinFlipper";

beforeEach(function() {
    jest
      .spyOn(Math, "random")
      .mockReturnValueOnce(0.25)
      .mockReturnValueOnce(0.75);
});

test('renders without crashing', () => {
    render(<CoinFlipper />);
});

test('matches snapshot', () => {
    const { asFragment } = render(<CoinFlipper />);
    expect(asFragment()).toMatchSnapshot();
  });
  

test('there is no image at the beginning', () => {
    const { queryByTestId } = render(<CoinFlipper />);
    expect(queryByTestId("coin-image")).toBeNull();
});

test('there is image show after clicking and it is head', () => {
    const { queryByTestId, queryByAltText, getByText } = render(<CoinFlipper />);

    const button = queryByTestId("flipping-button")
    fireEvent.click(button);

    expect(queryByTestId("coin-image")).toBeInTheDocument();

    // by mocking random we know first coin will be head
    expect(queryByAltText("head")).toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();
    expect(getByText("Out of 1, head: 1 and tail: 0")).toBeInTheDocument();
});

test('after clicking twice, check it is tail', () => {
    const { queryByTestId, queryByAltText, getByText } = render(<CoinFlipper />);

    const button = queryByTestId("flipping-button")
    fireEvent.click(button);
    fireEvent.click(button);

    expect(queryByTestId("coin-image")).toBeInTheDocument();

    // by mocking random we know first coin will be head
    expect(queryByAltText("tail")).toBeInTheDocument();
    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(getByText("Out of 2, head: 1 and tail: 1")).toBeInTheDocument();
});
  
afterEach(function() {
    Math.random.mockRestore();
});
