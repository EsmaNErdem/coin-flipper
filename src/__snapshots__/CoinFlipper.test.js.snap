// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinFlipper"
  >
    <h1>
      Let's flip
    </h1>
    <button
      data-testid="flipping-button"
    >
      Flip it!
    </button>
    <p>
       Out of 0, head: 0 and tail: 0
    </p>
  </div>
</DocumentFragment>
`;
